src/NServiceBus.Core/CriticalError/CriticalError.cs

```diff
@@ -49,17 +56,43 @@ public virtual void Raise(string errorMessage, Exception exception)
             Guard.AgainstNull(nameof(exception), exception);
             LogManager.GetLogger("NServiceBus").Fatal(errorMessage, exception);
 
-            if (Endpoint == null)
+            lock (endpointCriticalLock)
```

Comment:
There can be multiple threads raising a critical error. If the SetEndpoint method is called and the endpoint is set and the errors are being pulled from the shared list, we need to avoid contention.